Add paper list entry component

Merseleo opened the pull request:
Closes #101 

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a new button, which does x -->

- added a component for displaying the basic paper information and optionally review information about this paper
- added event handlers to this component to navigate to the corresponding paper view on double click and executing a custom function on single click (or navigate to the paper view, if no other functionality is desired)
- added a new npm command for running the tests and opening the coverage: `npm run show-coverage`
- outsource avatar from user menu to own component, that optionally can display the review decision of the corresponding user

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - ~~[ ] unit tests~~ (this PR adds a new component, which have no testable units)
  - [x] integration tests
  - ~~[ ] end-to-end tests~~ (this PR adds a new component, which cannot be tested e2e)
- [x] (for reviewer) I have checked the implementation against the requirements

Commit:
feat: add basic layout for paper entry component

Commit:
refactor: integrated required types for the paper entry component into model/backend.ts / adapted existing types and adapted related code

Commit:
feat: extended avatar component with indicator for review decision of the associated user and integrated the avatars in the paper entry component (+ changed userId to user in Review type, as userIds can not resolved to users yet)

Commit:
docs: add component comment for paper entry component (+ prepare onClick method of component)

Commit:
feat: add onclick event handling to paper entry component (either navigate to paper view page or execute a custom function, such as opening an overlay) and added second paper entry on component usage page to demonstrate this functionality

Commit:
chore: add command 'npm run show-coverage' to run the tests, build the coverage report and open it

Commit:
fix: do not call onclick function twice too, if double click event was triggered, but distinguish between single and double click

Commit:
test: add integration tests for the paper entry component (+ update search bar tests to use expect instead of assert)

Commit:
feat: change background color of paper entry component on hover

Slartibartfass2 commented on package.json:
```diff
@@ -11,6 +11,7 @@
         "format": "prettier --write .",
         "lint": "prettier --check . && eslint .",
         "test": "npm run test:unit && npm run test:integration && npm run test:e2e",
+        "show-coverage": "npm run test && cd coverage/ && open index.html",
```
How about:
```suggestion
        "show-coverage": "open coverage/index.html",
```
I often run only the integration test and then want to open the coverage, so I think we should have that separated. We could also add a command that combines both, e.g. `test:show-coverage`

Merseleo replied:
I think your second suggestion sounds like a good idea, I will add this command and adapt the first one.

Slartibartfass2 commented on src/lib/components/composites/navigation-bar/UserMenu.svelte:
```diff
@@ -52,11 +50,8 @@
-        </Avatar.Root>
-    </DropdownMenu.Trigger>
+    <UserAvatar {user} />
+    <DropdownMenu.Trigger></DropdownMenu.Trigger>
```
The avatar has to be in the trigger. Otherwise, it's not clickable.
```suggestion
    <DropdownMenu.Trigger>
        <UserAvatar {user} />
    </DropdownMenu.Trigger>
```

Merseleo replied:
Oh my fault, thank you for the hint.

Slartibartfass2 commented on src/lib/components/composites/user-avatar/UserAvatar.svelte:
```diff
@@ -0,0 +1,62 @@
+    import * as Avatar from "$lib/components/primitives/avatar/index.js";
+    import { ReviewDecision, type User } from "$lib/types";
+    import Check from "lucide-svelte/icons/check";
+    import { X } from "lucide-svelte";
```
I am not sure if this makes a difference:
```suggestion
    import X from "lucide-svelte/icons/x";
```

Merseleo replied:
I don't think so, but it looks better, so I will change it.

Slartibartfass2 commented on src/lib/components/composites/user-avatar/UserAvatar.svelte:
```diff
@@ -0,0 +1,62 @@
+    <Avatar.Root>
+        <Avatar.Fallback>{userInitials}</Avatar.Fallback>
+    </Avatar.Root>
+    {#if reviewDecision === ReviewDecision.Accepted}
+        <div
+            class="absolute top-6 left-6 flex justify-center items-center rounded-full bg-accept-green h-5 w-5"
+        >
+            <Check size={16} color="#ffffff" strokeWidth="3" />
+        </div>
+    {:else if reviewDecision === ReviewDecision.Declined}
+        <div
+            class="absolute top-6 left-6 flex justify-center items-center rounded-full bg-decline-red h-5 w-5"
+        >
+            <X size={16} color="#ffffff" strokeWidth="3" />
+        </div>
+    {:else if reviewDecision === ReviewDecision.Maybe}
+        <div
+            class="absolute top-6 left-6 flex justify-center items-center rounded-full bg-maybe-yellow h-5 w-5"
+        >
+            <text class="text-white">?</text>
+        </div>
+    {/if}
```
Do you think there's a way to modularize this a bit more? The divs all have the same style except the color.
I think this is optional, I once tried using some sort of config for the decision buttons but ended up in creating a component for each of them.

Merseleo replied:
I try to simplify it a bit, but I think an own component is not necessary as it is really simple and especially the icons change, so the only thing in common is the style.

Slartibartfass2 replied:
I think this is fine, but also add `h-5 w-5` to the style, which is the same for each div.

Merseleo replied:
Sorry, missed it.

Slartibartfass2 commented on tests/integration/navigation-bar/paper-navigation-bar.test.ts:
```diff
@@ -62,6 +62,7 @@ describe("PaperNavigationBar", () => {
                 },
                 backRef: "/",
                 paper: {
+                    id: undefined,
```
Was there a problem, when omitting the ID?

Merseleo replied:
No, if I remember correctly this was added while the rebase or so and I wanted to delete it, but forgot it, I will change it.

Slartibartfass2 commented on src/routes/+page.svelte:
```diff
@@ -8,4 +8,6 @@
     <title>SnowballR</title>
 </svelte:head>
-<SimpleNavigationBar {user} title="SnowballR" tabs={[]} defaultTabValue="" />
+<div class="flex flex-col gap-3">
+    <SimpleNavigationBar {user} title="SnowballR" tabs={[]} defaultTabValue="" />
+</div>
```
What was the reason for this change?

Merseleo replied:
Because first I tested the components on this page and therefore wanted this layout, but if I am not mistaken, this will be changed after the rebase anyway?

Slartibartfass2 replied:
I'm not aware of any changes. In my PR of the Paper View (#89) I only added styling to the global container in the +layout.svelte.

Merseleo replied:
Ah, you are right. The styling is in the +layout.svelte. I will delete it.

Slartibartfass2 commented on src/lib/components/composites/paper-components/PaperEntry.svelte:
```diff
@@ -0,0 +1,91 @@
+        [ReviewDecision.Declined]: "border-decline-red",
+    };
+
+    const showStatus = () => showReviewStatus && paper.reviewData !== undefined;
```
Why does this have to be a method and not just a boolean value?
```suggestion
    const isStatusVisible = showReviewStatus && paper.reviewData !== undefined;
```

Merseleo replied:
It doesn't have to, I changed it.

Slartibartfass2 commented on src/lib/components/composites/paper-components/PaperEntry.svelte:
```diff
@@ -0,0 +1,91 @@
+            if (timeoutId !== null) clearTimeout(timeoutId);
+            navigateToPaperView();
+        }
+    };
```
This method looks a bit confusing at first. I think we can make it more self-describing.
```suggestion
    const handleClick = (event: MouseEvent) => {
        if (timeoutId !== null) {
            clearTimeout(timeoutId);
        }
        const numberOfClicks = event.detail;

        if (numberOfClicks === 1) {
            timeoutId = setTimeout(onClick, 250);
        } else if (numberOfClicks === 2) {
            navigateToPaperView();
        }
    };
```
- renamed `e` to `event`
- moved `timeoutId != null` check to the beginning
  - check is made for single and double click
  - else case isn't reached when we navigate to the paper view after a double click
- introduced new variable `numberOfClick` to make the meaning of `event.detail` more clear
- `() => onClick()` can be just `onClick` (the former just wraps `onClick` in another method)

What are your thoughts on these changes?

Merseleo replied:
I think your code is more elegant and I would change the timeout to 500ms, as you proposed (and Microsoft uses).
250ms was only a first try to make it faster, but I agree that the interval between the two clicks for a double click could be too short.

Slartibartfass2 replied:
I took a second look at this today and found some weird race conditions. If you click a second time, just when the timer stops, both events are fired. This is easy to achieve with a high cooldown, like 1 second.

I think this is due to the different interpretation of `numberOfClicks` and that the `timeoutId` isn't reset.

I investigated this for a while and found this, to be a sound solution:
- if the timeout wasn't set yet, it's the first click
  - start the timeout
  - when the timeout finishes, the `timeoutId` will be reset and the single click event fired
- when the timeout was set, it's the second click
  - clear the running timeout
  - reset `timoutId`
  - fire double click event
```ts
    const handleClick = () => {
        if (timeoutId === null) {
            timeoutId = setTimeout(() => {
                timeoutId = null;
                onClick();
            }, 350);
        } else {
            clearTimeout(timeoutId);
            timeoutId = null;
            navigateToPaperView();
        }
    };
```

What are your thoughts on this? Could you reproduce the behavior, which I described?

Also, on a second thought, 500 ms is too much, the single click has a high delay. We could do maybe 350ms?

Lastly, don't forget to adjust the documentation, there's a different amount of delay in it.

Merseleo replied:
I am also fine with 350ms, I also experimented with the delay an 250ms was the best for me, but I understand that it could be too fast for some people and we should increase it. Thus, i will change it (and update the documentation ;))

Slartibartfass2 commented on src/lib/components/composites/paper-components/PaperEntry.svelte:
```diff
@@ -0,0 +1,91 @@
+    type="button"
+    class="flex flex-row items-center justify-end pe-4 gap-3 w-full border {showStatus()
+        ? 'border-l-0'
+        : ''} border-container-border-grey rounded-md hover:bg-container-border-grey/35"
```
Optionally, you could use the `cn` utils method here to add additional style.
There's a similar case below.
```suggestion
    class={cn(
        "flex flex-row items-center justify-end pe-4 gap-3 w-full border border-container-border-grey rounded-md hover:bg-container-border-grey/35",
        isStatusVisible ? "border-l-0" : "",
    )}
```

Merseleo replied:
Theoretically I could and used It at the beginning, but personally I like the the current style more. 
But if you want to make it consistent, I can change it?

Slartibartfass2 replied:
No, that's up to you.

Slartibartfass2 commented on src/lib/components/composites/paper-components/PaperEntry.svelte:
```diff
@@ -0,0 +1,91 @@
+        <PaperInfo {paper} />
+    </div>
+    {#if showStatus()}
+        {#each paper.reviewData!.individualDecisions as [user, individualDecision]}
```
The `!` is not necessary here because TypeScript checks that `reviewData` is defined.
Alternatively, you can remove the `paper.reviewData !== undefined` check and use `paper.reviewData?.individualDecisions ?? []`.

Afaik, `!` is a bad practice because you fool the compiler to believe that something is defined.

Merseleo replied:
As showStatus() was a method, I got an error or at least a hint, that paper.reviewData could be null so I needed to add this check (or I was lost). But as the check is made via a variable now, it is completely fine to drop the ! and stick with best practice and not the bad practice.

Slartibartfass2 replied:
great!

Commit:
refactor: renamed paper entry component to paper list entry component

Slartibartfass2 reviewed (changes requested):
The component looks great, nice job!

I still have two questions:
1. Should we increase the double click cooldown? There's no standardization, so this leaves us to decide what's best. [Microsoft](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setdoubleclicktime#parameters) uses 500ms as default and I think that's reasonable, but I'm open to discussion.
2. When the component is hovered, there's little contrast between the hovered component and the avatar, do you think we should change this?

Commit:
chore: rename command to run the tests and open the coverage to 'test:and-open-coverage' and add a second command 'show-coverage' to only open the coverage report

Merseleo committed:
refactor: apply suggestions from review

This includes deleting the unnecessary paper id in the paper-navigation bar test, changing the import for the X icon and integrating the UserAvatar component correctly, so the UserMenu can be opened.

Co-authored-by: Felix Schlegel <40503329+Slartibartfass2@users.noreply.github.com>

Commit:
refactor: improve structure of the handleClick method and increase timeout interval for second click to 500ms

Merseleo reviewed (commented):
Thank you for this great review, some very good catches and nice feedback.

Regarding your second question: I agree, that it is not the best solution, but I wanted to wait for your feedback first and see if you had any ideas. However, I will add my idea and you can give feedback on it.

Hopefully, I didn't forgot anything.

P.S. I will rebase, after we completed this review and will re-request a review from you, but I think there will not be any major changes.

Commit:
refactor: change style of user avatar (make background white on hover of paper entry component) and combine div styling of review decision variants into a single class

Slartibartfass2 commented on src/lib/components/composites/user-avatar/UserAvatar.svelte:
```diff
@@ -0,0 +1,64 @@
+-->
+<div class="relative">
+    <Avatar.Root>
+        <Avatar.Fallback class="group-hover/paper-list-entry:bg-white"
+            >{userInitials}</Avatar.Fallback
+        >
```
Minor, but this weirds me out 😅 
```suggestion
        <Avatar.Fallback class="group-hover/paper-list-entry:bg-white">
            {userInitials}
        </Avatar.Fallback>
```

Merseleo replied:
Me too, I changed it.

Slartibartfass2 reviewed (changes requested):
This is nearly done! Great job.

Slartibartfass2 commented:
To your hover style change:
What do you think about `bg-slate-200`, this makes everything a bit darker and keeps the contrast.

Commit:
refactor: apply suggestions from review (no styling in global +page.svelte, as this will be in layout, change background color of user avatar on hover and outsource size into correspondign tailwind class

Merseleo reviewed (commented):
Regarding your proposal to change the background of the user avatar to `bg-slate-200` on hover looks better, thank you!

Commit:
fix: single click event not fired anymore, when a double click event happened (with a certain time interval between the clicks)

Slartibartfass2 reviewed (changes requested):
Let's rebase and finish this!
